pre-sync/oci-image-verification/main.go

```diff
@@ -0,0 +1,145 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os/exec"
+	"sync"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	SourceURL    = "configsync.gke.io/source-url"
+	SourceCommit = "configsync.gke.io/source-commit"
+)
+
+var authorized bool
+var authMutex sync.Mutex // Add a mutex for thread safety
+
+// Function to extract annotations from JSON
+func getAnnotations(raw []byte) (map[string]string, error) {
+	var metadata map[string]interface{}
+	if err := json.Unmarshal(raw, &metadata); err != nil {
+		return nil, err
+	}
+	if annotations, ok := metadata["metadata"].(map[string]interface{})["annotations"].(map[string]interface{}); ok {
+		annotationsMap := make(map[string]string)
+		for k, v := range annotations {
+			annotationsMap[k] = fmt.Sprintf("%v", v)
+		}
+		return annotationsMap, nil
+	}
+	return nil, fmt.Errorf("no annotations found")
+}
+
+func validateImage(image string) error {
+	cmd := exec.Command("cosign", "verify", image, "--key", "/cosign-key/cosign.pub")
+	log.Printf("command %s", cmd.String())
+	output, err := cmd.CombinedOutput()
+	if err != nil {
```

Comment:
Do we need to check the output content? I'm wondering if it's possible that err is nil, but output indicates a verification error.

Reply:
The current validation is not sufficient. It doesn't address the "time of check" and "time of use" race condition. The verified image digest may not match the new fetched digest in the `source-commit` annotation. We need to pass the source-commit and check if the digest match.

The verified digest can be retrieved from `cosign verify` output.

Reply:
The error can be captured with the CombinedOutput